### DISCUSSION TITLE
HOGWARTS PROJECT: my forEach loop errors Uncaught TypeError: hatArr.forEach is not a function

### DISCUSSION THREAD

**Discussion by cmeffley** [nss-evening-cohort-14/evening-client-side#68] [Help Me!] [answered]
Chris Meffley

Did not see any similar discussion tickets

When I submit my value in my form, I get the error message:
 Uncaught TypeError: hatArr.forEach is not a function
    at HTMLButtonElement.cardCreated 

https://stackoverflow.com/questions/31096596/why-is-foreach-not-a-function-for-this-object/31096661
https://stackoverflow.com/questions/60823400/uncaught-typeerror-foreach-is-not-a-function-error-in-javascript-function
There doesn't seem to be many helpful links. Other one's I've found talk about JSON, how a for loop is better than a forEach, or people confusing strings.

I haven't tried much, because I'm not sure what else to do. My syntax seems to be ok.

Link:
https://github.com/cmeffley/sorting-hat/blob/form/main.js
Ignore the createForm portion. I've senced fixed it but not pushed.

**Comment by TrinityChristiana** [comment]
Try console logging what `hatArr` is and check it's data type. That should give you a hit as to why it's throwing you an error. Report back if this doesn't help 😄

**Comment by cmeffley** [comment, marked as answer]
In my function below, I had a parameter defined that did not need to be defined cardCreated (hatArr), should be cardCreated( ). Then I had hatArr.forEach, but since I did not need to define the parameter, I needed to pass the array that the function needed to loop through. So, for this, I changed it to hat.forEach
![2021-02-01](https://user-images.githubusercontent.com/76795299/106547091-8b70b680-64da-11eb-97b4-fe45e405b7d4.png)

OPTION: if I wanted to keep the hatArr in the parameter and in front of the forEach, I could have done that, IF I  made an anonymous function where I called cardCreated. So it would have looked like this:
```
document.querySelector(#'sort1).addEventListener('click', () => {
    cardCreated(hat);
});
```

The empty `() =>` makes an anonymous function and cardCreated is calling my hat array to loop through the function.
